core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -322,51 +327,81 @@ public void close()
             implements TaskSource
     {
         private final Map<PlanNodeId, SplitSource> splitSources;
+        private final IdentityHashMap<ExchangeSourceHandle, Exchange> exchangeForHandle;
         private final Multimap<PlanNodeId, ExchangeSourceHandle> partitionedExchangeSourceHandles;
         private final Multimap<PlanNodeId, ExchangeSourceHandle> replicatedExchangeSourceHandles;
+
         private final int splitBatchSize;
         private final LongConsumer getSplitTimeRecorder;
         private final int[] bucketToPartitionMap;
         private final Optional<BucketNodeMap> bucketNodeMap;
         private final Optional<CatalogName> catalogRequirement;
+        private final long targetPartitionSourceSizeInBytes; // compared data read from ExchangeSources
+        private final long targetPartitionSplitWeight; // compared against splits from SplitSources
 
-        private boolean finished;
+        private volatile boolean finished;
```

Comment:
nit: does this have to be volatile? The `TaskSource` is not expected to be thread safe and should always be accessed from a single thread